### PR TITLE
Implement thamos graph command

### PR DESCRIPTION
**Pull request by fridex**
## Related Issues and Dependencies

Related: https://github.com/thoth-station/adviser/issues/1978

## This introduces a breaking change

- [x] No

### CODE REVIEW

**Comment by fridex** [comment]
Shared dependency sub-graphs are referenced using "…":

```
$ thamos graph
→ numpy==1.19.3 from https://pypi.org/simple
→ thoth-common==0.20.4 from https://pypi.org/simple
  ↳ openshift==0.11.2 from https://pypi.org/simple
    ↳ kubernetes==11.0.0 from https://pypi.org/simple
      ↳ urllib3==1.25.10 from https://pypi.org/simple
      ↳ six==1.15.0 from https://pypi.org/simple
      ↳ requests-oauthlib==1.3.0 from https://pypi.org/simple
        ↳ requests==2.24.0 from https://pypi.org/simple
          ↳ chardet==3.0.4 from https://pypi.org/simple
          ↳ certifi==2020.6.20 from https://pypi.org/simple
          ↳ urllib3==1.25.10 from https://pypi.org/simple
            ↳ …
          ↳ idna==2.10 from https://pypi.org/simple
        ↳ oauthlib==3.1.0 from https://pypi.org/simple
      ↳ setuptools==50.3.0 from https://pypi.org/simple
      ↳ pyyaml==5.3.1 from https://pypi.org/simple
      ↳ python-dateutil==2.8.1 from https://pypi.org/simple
        ↳ six==1.15.0 from https://pypi.org/simple
          ↳ …
      ↳ certifi==2020.6.20 from https://pypi.org/simple
        ↳ …
      ↳ requests==2.24.0 from https://pypi.org/simple
        ↳ …
      ↳ websocket-client==0.57.0 from https://pypi.org/simple
        ↳ six==1.15.0 from https://pypi.org/simple
          ↳ …
      ↳ google-auth==1.21.3 from https://pypi.org/simple
        ↳ setuptools==50.3.0 from https://pypi.org/simple
          ↳ …
        ↳ rsa==4.6 from https://pypi.org/simple
          ↳ pyasn1==0.4.8 from https://pypi.org/simple
        ↳ six==1.15.0 from https://pypi.org/simple
          ↳ …
        ↳ pyasn1-modules==0.2.8 from https://pypi.org/simple
          ↳ pyasn1==0.4.8 from https://pypi.org/simple
            ↳ …
        ↳ cachetools==4.1.1 from https://pypi.org/simple
    ↳ jinja2==2.11.2 from https://pypi.org/simple
      ↳ markupsafe==1.1.1 from https://pypi.org/simple
    ↳ six==1.15.0 from https://pypi.org/simple
      ↳ …
    ↳ python-string-utils==1.0.0 from https://pypi.org/simple
    ↳ ruamel-yaml==0.16.12 from https://pypi.org/simple
      ↳ ruamel-yaml-clib==0.2.2 from https://pypi.org/simple
  ↳ daiquiri==2.1.1 from https://pypi.org/simple
    ↳ python-json-logger==0.1.11 from https://pypi.org/simple
  ↳ kubernetes==11.0.0 from https://pypi.org/simple
    ↳ …
  ↳ argo-workflows==3.5.1 from https://pypi.org/simple
    ↳ mock==4.0.2 from https://pypi.org/simple
    ↳ kubernetes==11.0.0 from https://pypi.org/simple
      ↳ …
  ↳ attrdict==2.0.1 from https://pypi.org/simple
    ↳ six==1.15.0 from https://pypi.org/simple
      ↳ …
  ↳ attrs==20.2.0 from https://pypi.org/simple
  ↳ rfc5424-logging-handler==1.4.3 from https://pypi.org/simple
    ↳ pytz==2020.1 from https://pypi.org/simple
    ↳ tzlocal==2.1 from https://pypi.org/simple
      ↳ pytz==2020.1 from https://pypi.org/simple
        ↳ …
  ↳ thoth-common==0.20.4 from https://pypi.org/simple
    ↳ …
  ↳ pyyaml==5.3.1 from https://pypi.org/simple
    ↳ …
  ↳ jsonformatter==0.3.0 from https://pypi.org/simple
  ↳ requests==2.24.0 from https://pypi.org/simple
    ↳ …
  ↳ sentry-sdk==0.19.1 from https://pypi.org/simple
    ↳ certifi==2020.6.20 from https://pypi.org/simple
      ↳ …
    ↳ urllib3==1.25.10 from https://pypi.org/simple
      ↳ …
```

File: thamos/lib.py [hunk 1]
```diff
@@ -1010,8 +1011,16 @@ def get_status(api_client: ApiClient, analysis_id: str = None):
 
 
 @with_api_client
-def get_analysis_results(api_client: ApiClient, analysis_id: str):
+def get_analysis_results(
+    api_client: ApiClient, analysis_id: typing.Optional[str] = None
+):
     """Get the analysis result from a given id."""
+    with open("/home/fpokorny/git/thoth-station/adviser/adviser_document.json") as f:
```
Comment (pacospace) [inline thamos/lib.py:1018]: it should be an env variable for path?

Comment (fridex) [reply thamos/lib.py]: That was a file used to run experiments. Removed. Thanks for spotting it 👍🏻

**Comment by goern** [comment]
@Gkrumbach07 what do you think about adding a machine readable aka json output for the graph?

**Comment by Gkrumbach07** [comment]
> @Gkrumbach07 what do you think about adding a machine readable aka json output for the graph?

That would be preferable. Currently I am putting it into the form `parent : [child, child, child]`. So in the case of python packages, the graph is a js map object with the names of packages as keys and the value as an array of dependency keys.

**Comment by goern** [comment]
@fridex uhm, maybe I am wrong, as the JSON is delivered by the User API anyways? right?

**Comment by fridex** [comment]
> @fridex uhm, maybe I am wrong, as the JSON is delivered by the User API anyways? right?

Yes. However, the output can be also read using `thamos advise --no-write --json` if needed.

**Comment by fridex** [comment]
> > @Gkrumbach07 what do you think about adding a machine readable aka json output for the graph?
> 
> That would be preferable. Currently I am putting it into the form `parent : [child, child, child]`. So in the case of python packages, the graph is a js map object with the names of packages as keys and the value as an array of dependency keys.

The backend serializes output as nodes (stating package names) and edges between nodes encoded as (source_node, destination_node). Nodes are referenced as indexes to the nodes listing.

**Comment by fridex** [comment]
Added `thamos graph --unfold` to unfold collapsed shared sub-graphs. This output is more verbose but states the whole dependency graph to terminal (except cyclic dependendencies that are marked with three dots).

Example of an unfolded dependency graph from stack shown above:

```
→ numpy==1.19.3 from https://pypi.org/simple
→ thoth-common==0.20.4 from https://pypi.org/simple
  ↳ openshift==0.11.2 from https://pypi.org/simple
    ↳ kubernetes==11.0.0 from https://pypi.org/simple
      ↳ urllib3==1.25.10 from https://pypi.org/simple
      ↳ six==1.15.0 from https://pypi.org/simple
      ↳ requests-oauthlib==1.3.0 from https://pypi.org/simple
        ↳ requests==2.24.0 from https://pypi.org/simple
          ↳ chardet==3.0.4 from https://pypi.org/simple
          ↳ certifi==2020.6.20 from https://pypi.org/simple
          ↳ urllib3==1.25.10 from https://pypi.org/simple
          ↳ idna==2.10 from https://pypi.org/simple
        ↳ oauthlib==3.1.0 from https://pypi.org/simple
      ↳ setuptools==50.3.0 from https://pypi.org/simple
      ↳ pyyaml==5.3.1 from https://pypi.org/simple
      ↳ python-dateutil==2.8.1 from https://pypi.org/simple
        ↳ six==1.15.0 from https://pypi.org/simple
      ↳ certifi==2020.6.20 from https://pypi.org/simple
      ↳ requests==2.24.0 from https://pypi.org/simple
        ↳ chardet==3.0.4 from https://pypi.org/simple
        ↳ certifi==2020.6.20 from https://pypi.org/simple
        ↳ urllib3==1.25.10 from https://pypi.org/simple
        ↳ idna==2.10 from https://pypi.org/simple
      ↳ websocket-client==0.57.0 from https://pypi.org/simple
        ↳ six==1.15.0 from https://pypi.org/simple
      ↳ google-auth==1.21.3 from https://pypi.org/simple
        ↳ setuptools==50.3.0 from https://pypi.org/simple
        ↳ rsa==4.6 from https://pypi.org/simple
          ↳ pyasn1==0.4.8 from https://pypi.org/simple
        ↳ six==1.15.0 from https://pypi.org/simple
        ↳ pyasn1-modules==0.2.8 from https://pypi.org/simple
          ↳ pyasn1==0.4.8 from https://pypi.org/simple
        ↳ cachetools==4.1.1 from https://pypi.org/simple
    ↳ jinja2==2.11.2 from https://pypi.org/simple
      ↳ markupsafe==1.1.1 from https://pypi.org/simple
    ↳ six==1.15.0 from https://pypi.org/simple
    ↳ python-string-utils==1.0.0 from https://pypi.org/simple
    ↳ ruamel-yaml==0.16.12 from https://pypi.org/simple
      ↳ ruamel-yaml-clib==0.2.2 from https://pypi.org/simple
  ↳ daiquiri==2.1.1 from https://pypi.org/simple
    ↳ python-json-logger==0.1.11 from https://pypi.org/simple
  ↳ kubernetes==11.0.0 from https://pypi.org/simple
    ↳ urllib3==1.25.10 from https://pypi.org/simple
    ↳ six==1.15.0 from https://pypi.org/simple
    ↳ requests-oauthlib==1.3.0 from https://pypi.org/simple
      ↳ requests==2.24.0 from https://pypi.org/simple
        ↳ chardet==3.0.4 from https://pypi.org/simple
        ↳ certifi==2020.6.20 from https://pypi.org/simple
        ↳ urllib3==1.25.10 from https://pypi.org/simple
        ↳ idna==2.10 from https://pypi.org/simple
      ↳ oauthlib==3.1.0 from https://pypi.org/simple
    ↳ setuptools==50.3.0 from https://pypi.org/simple
    ↳ pyyaml==5.3.1 from https://pypi.org/simple
    ↳ python-dateutil==2.8.1 from https://pypi.org/simple
      ↳ six==1.15.0 from https://pypi.org/simple
    ↳ certifi==2020.6.20 from https://pypi.org/simple
    ↳ requests==2.24.0 from https://pypi.org/simple
      ↳ chardet==3.0.4 from https://pypi.org/simple
      ↳ certifi==2020.6.20 from https://pypi.org/simple
      ↳ urllib3==1.25.10 from https://pypi.org/simple
      ↳ idna==2.10 from https://pypi.org/simple
    ↳ websocket-client==0.57.0 from https://pypi.org/simple
      ↳ six==1.15.0 from https://pypi.org/simple
    ↳ google-auth==1.21.3 from https://pypi.org/simple
      ↳ setuptools==50.3.0 from https://pypi.org/simple
      ↳ rsa==4.6 from https://pypi.org/simple
        ↳ pyasn1==0.4.8 from https://pypi.org/simple
      ↳ six==1.15.0 from https://pypi.org/simple
      ↳ pyasn1-modules==0.2.8 from https://pypi.org/simple
        ↳ pyasn1==0.4.8 from https://pypi.org/simple
      ↳ cachetools==4.1.1 from https://pypi.org/simple
  ↳ argo-workflows==3.5.1 from https://pypi.org/simple
    ↳ mock==4.0.2 from https://pypi.org/simple
    ↳ kubernetes==11.0.0 from https://pypi.org/simple
      ↳ urllib3==1.25.10 from https://pypi.org/simple
      ↳ six==1.15.0 from https://pypi.org/simple
      ↳ requests-oauthlib==1.3.0 from https://pypi.org/simple
        ↳ requests==2.24.0 from https://pypi.org/simple
          ↳ chardet==3.0.4 from https://pypi.org/simple
          ↳ certifi==2020.6.20 from https://pypi.org/simple
          ↳ urllib3==1.25.10 from https://pypi.org/simple
          ↳ idna==2.10 from https://pypi.org/simple
        ↳ oauthlib==3.1.0 from https://pypi.org/simple
      ↳ setuptools==50.3.0 from https://pypi.org/simple
      ↳ pyyaml==5.3.1 from https://pypi.org/simple
      ↳ python-dateutil==2.8.1 from https://pypi.org/simple
        ↳ six==1.15.0 from https://pypi.org/simple
      ↳ certifi==2020.6.20 from https://pypi.org/simple
      ↳ requests==2.24.0 from https://pypi.org/simple
        ↳ chardet==3.0.4 from https://pypi.org/simple
        ↳ certifi==2020.6.20 from https://pypi.org/simple
        ↳ urllib3==1.25.10 from https://pypi.org/simple
        ↳ idna==2.10 from https://pypi.org/simple
      ↳ websocket-client==0.57.0 from https://pypi.org/simple
        ↳ six==1.15.0 from https://pypi.org/simple
      ↳ google-auth==1.21.3 from https://pypi.org/simple
        ↳ setuptools==50.3.0 from https://pypi.org/simple
        ↳ rsa==4.6 from https://pypi.org/simple
          ↳ pyasn1==0.4.8 from https://pypi.org/simple
        ↳ six==1.15.0 from https://pypi.org/simple
        ↳ pyasn1-modules==0.2.8 from https://pypi.org/simple
          ↳ pyasn1==0.4.8 from https://pypi.org/simple
        ↳ cachetools==4.1.1 from https://pypi.org/simple
  ↳ attrdict==2.0.1 from https://pypi.org/simple
    ↳ six==1.15.0 from https://pypi.org/simple
  ↳ attrs==20.2.0 from https://pypi.org/simple
  ↳ rfc5424-logging-handler==1.4.3 from https://pypi.org/simple
    ↳ pytz==2020.1 from https://pypi.org/simple
    ↳ tzlocal==2.1 from https://pypi.org/simple
      ↳ pytz==2020.1 from https://pypi.org/simple
  ↳ thoth-common==0.20.4 from https://pypi.org/simple
    ↳ …
  ↳ pyyaml==5.3.1 from https://pypi.org/simple
  ↳ jsonformatter==0.3.0 from https://pypi.org/simple
  ↳ requests==2.24.0 from https://pypi.org/simple
    ↳ chardet==3.0.4 from https://pypi.org/simple
    ↳ certifi==2020.6.20 from https://pypi.org/simple
    ↳ urllib3==1.25.10 from https://pypi.org/simple
    ↳ idna==2.10 from https://pypi.org/simple
  ↳ sentry-sdk==0.19.1 from https://pypi.org/simple
    ↳ certifi==2020.6.20 from https://pypi.org/simple
    ↳ urllib3==1.25.10 from https://pypi.org/simple
```

Ready to be reviewed.

**Comment by fridex** [comment]
Will merge this to continue with work. Feel free to raise any comments/concerns/suggestions.

**Pull request merged**